src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,642 @@
+        }
+
+        /// <inheritDoc/>
+        public bool TryGetValue(object key, out object value)
+        {
+            if (!IsEmpty)
+            {
+                PropertyValue valueItem;
+                if (EventProperties.TryGetValue(key, out valueItem))
+                {
+                    value = valueItem.Value;
+                    return true;
+                }
+            }
+
+            value = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Attempt to use the message-template-parameters without allocating a dictionary
+        /// </summary>
+        /// <param name="parameterList">Message-template-parameters</param>
+        /// <returns>List of message-template-parameters if succesful (else null)</returns>
+        static IList<MessageTemplateParameter> CreateUniqueMessagePropertiesListFast(IList<MessageTemplateParameter> parameterList)
+        {
+            if (parameterList.Count <= 10)
+            {
+                bool uniqueMessageProperties = true;
+                for (int i = 0; i < parameterList.Count - 1; ++i)
+                {
+                    for (int j = i + 1; j < parameterList.Count; ++j)
+                    {
+                        if (parameterList[i].Name == parameterList[j].Name)
+                        {
+                            uniqueMessageProperties = false;
+                            break;
+                        }
+                    }
+                }
+                if (uniqueMessageProperties)
+                {
+                    var messageProperties = new MessageTemplateParameter[parameterList.Count];
+                    for (int i = 0; i < parameterList.Count; ++i)
+                        messageProperties[i] = parameterList[i];
+                    return messageProperties;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Attempt to insert the message-template-parameters into an empty dictionary
+        /// </summary>
+        /// <param name="messageProperties">Message-template-parameters</param>
+        /// <param name="eventProperties">The initially empty dictionary</param>
+        /// <returns>Message-template-parameters was inserted into dictionary without trouble (true/false)</returns>
+        private static bool InsertMessagePropertiesIntoEmptyDictionary(IList<MessageTemplateParameter> messageProperties, Dictionary<object, PropertyValue> eventProperties)
+        {
+            try
+            {
+                for (int i = 0; i < messageProperties.Count; ++i)
+                {
+                    eventProperties.Add(messageProperties[i].Name, new PropertyValue(messageProperties[i].Value, true));
+                }
+                return true; // We are done
+            }
+            catch (ArgumentException)
+            {
+                // Duplicate keys found, lets try again
+                for (int i = 0; i < messageProperties.Count; ++i)
```

Comment:
Could you explain me how remove is a try-again for add?

Reply:
Have fixed the comment